raven/utilities/forecasting.py

```diff
@@ -0,0 +1,151 @@
+"""
+Tools for hydrological forecasting
+"""
+
+from raven.models import get_model
+import logging
+import xarray as xr
+import numpy as np
+import pandas as pd
+import datetime as dt
+import tempfile
+import pdb
+
+LOGGER = logging.getLogger("PYWPS")
+
+# This function gets model states after running the model (i.e. states at the end of the run).
+def getRavenStates(model, **kwds):
+    """Perform regionalization for catchment whose outlet is defined by coordinates.
+
+    Parameters
+    ----------
+    model : {'HMETS', 'GR4JCN', 'MOHYSE', 'HBVEC'}
+      Model name.
+    kwds : {}
+      Model configuration parameters, including the forcing files (ts).
+
+    Returns
+    -------
+    rvc : {}
+      Raven model forcing file
+        
+    """
+    # Run the model and get the rvc file for future hotstart.
+    m = get_model(model)()
+    m(overwrite=True, **kwds)
+    rvc = m.outputs['solution']
+   
+    return rvc
+
+
+'''
+Function that might be useful eventually to do a forecast from a model setup.
+'''
+# Do the actual forecasting step
+def performForecastingStep(rvc,model,**kwds):
+    # kwds includes 'ts', the forecast timeseries data
+    # Setup the model
+    m = get_model(model)()
+    
+    # Force the initial conditions
+    m.resume(rvc) 
+    
+    # Set the parameters, start dates, etc. required to run the model and run
+    m(overwrite=True, **kwds)
+
+    return m.q_sim
+  
+    
+    
+    
+    
+''' 
+This function takes the model setup and name as well as forecast data and lead time and returns
+an ESP forecast netcdf. The data comes from the climatology data and thus there is a mechanism
+to get the correct data from the time series and exclude the current year.
+'''
+def performClimatologyESP(model_name,ts,forecast_date, lead_time, **kwds):
+    
+    # Get the timeseries
+    tsnc=xr.open_dataset(ts)
+    
+    # Prepare model instance
+    m = get_model(model_name)()
+    
+    # Now find the periods of time for warm-up and forecast and add to the model keywords as the defaults are failing (nanoseconds datetimes do not like the year 0001...)
+    start_date=pd.to_datetime(tsnc['time'][0].values)
+    start_date=start_date.to_pydatetime()
+    kwds['start_date']=start_date
+   
+    # Check to make sure forecast date is not in the first year as we need model warm-up.
+    dateLimit = start_date.replace(year=start_date.year + 1)
+    if (forecast_date.month==2 and forecast_date.day==29):
+        forecast_date.replace(day=28)
```

Comment:
I thought about it and the problem is that if the first year is a leap year, adding 365 will bring us to Dec-31st instead of Jan-01 of next year. The odds of that being a problem are low but the solution to be more precise (using replace with a relative time) is not expensive at all so I'd leave it like this.